ESQL: Clarify docs and add csv test for WHERE in STATS

alex-spies opened the pull request.

alex-spies committed:
Clarify docs and add csv test for WHERE in STATS

bpintea commented on docs/reference/query-languages/esql/_snippets/commands/layout/stats-by.md:
```diff
@@ -31,7 +31,9 @@ STATS [column1 =] expression1 [WHERE boolean_expression1][,
 `boolean_expressionX`
-:   The condition that must be met for a row to be included in the evaluation of `expressionX`.
+:   The condition that must be met for a row to be included in the evaluation of
+    `expressionX`. Does not have any effect on the values used in
```
```suggestion
    `expressionX`. Does not have any effect on the values produced by
```

bpintea commented on docs/reference/query-languages/esql/_snippets/commands/layout/stats-by.md:
```diff
@@ -31,7 +31,9 @@ STATS [column1 =] expression1 [WHERE boolean_expression1][,
-:   The condition that must be met for a row to be included in the evaluation of `expressionX`.
+:   The condition that must be met for a row to be included in the evaluation of
+    `expressionX`. Does not have any effect on the values used in
+    `grouping_expressionX` or the other `expressionX`.
```
```suggestion
    `grouping_expressionX` or the other `expressionX`. Consequently, `... | STATS ... WHERE <condition> ...` is not equivalent to `... | WHERE <condition> | STATS ...`.
```

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -2420,6 +2420,43 @@ under_40K:long |inbetween:long |over_60K:long  |total:long
+// tag::aggFilteringOnGroup[]
+FROM employees
+| STATS `count` = COUNT(*) WHERE languages < 3 BY languages
+// end::aggFilteringNoGroup[]
```
```suggestion
// end::aggFilteringOnGroup[]
```

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -2420,6 +2420,43 @@ under_40K:long |inbetween:long |over_60K:long  |total:long
+
+docsStatsWithFilteringBefore
+required_capability: per_agg_filtering
+// tag::aggFilteringOnGroup[]
```
```suggestion
// tag::aggFilteringBefore[]
```

alex-spies committed:
Apply suggestions from code review

Co-authored-by: Bogdan Pintea <sig11@mailbox.org>

alex-spies committed:
Merge branch 'main' into docs-stats-where-groupby

leemthompo commented:
@alex-spies this seems to have fallen through the cracks, should this get merged? :)

alex-spies committed:
Merge branch 'main' into docs-stats-where-groupby

alex-spies committed:
Include new doc tests into example descriptions

alex-spies committed:
Merge branch 'main' into docs-stats-where-groupby

alex-spies commented:
Hey @leemthompo , I hooked in the new examples into `stats-by.md`. This is ready to go from my side, but would appreciate a quick :heavy_check_mark: from you before merging this :)

leemthompo committed:
copyedit and restructure

- `**syntax**`, `**parameters**`, `**description**` → `## h2` headings
- `### examples` → `## examples`
- `####` subsections promoted to `###`
- `#### naming` renamed to `### omitting column names`
- each example gets a concise `###` heading with a rewritten lead-in that adds context beyond the heading
- `boolean_expressionx` description: tightened prose, equivalence contrast moved to two fenced `esql` blocks

leemthompo committed:
fix wording

leemthompo committed:
add lead-in prose, fix example heading

alex-spies commented on docs/reference/query-languages/esql/_snippets/commands/layout/stats-by.md:
```diff
@@ -114,19 +137,23 @@ the aggregation:
 
-Compare this to filtering with `WHERE` before `STATS`, which excludes
-non-matching groups entirely:
+### WHERE before STATS excludes non-matching groups
```
nit: I think the following example is still part of `Filter on the grouping key` (by contrasting the example above). But this is fine, too.

alex-spies reviewed (commented):
Thanks @leemthompo , I think the new format is much better than before.

I have one nit, but we can also merge this as-is.

Let me know what you think and let's get this in :)

leemthompo committed:
fix filteron group section
